File: pkg/mu/bodies.go
package mu

import (
	"errors"
	"fmt"
)

// SessionAcquireBody is the payload for session.acquire.
type SessionAcquireBody struct {
	TTLMS int64 `json:"ttlMs"`
}

// SessionRenewBody is the payload for session.renew.
type SessionRenewBody struct {
	TTLMS int64 `json:"ttlMs"`
}

// SessionReplyBody is returned by session.acquire.
type SessionReplyBody struct {
	Session      SessionLease `json:"session"`
	StateVersion int64        `json:"stateVersion"`
}

// SessionLease describes a session lease response.
type SessionLease struct {
	ID             string `json:"id"`
	Token          string `json:"token"`
	Owner          string `json:"owner"`
	LeaseExpiresAt int64  `json:"leaseExpiresAt"`
}

// PlaybackPlayBody is the payload for playback.play.
type PlaybackPlayBody struct {
	Index *int64 `json:"index,omitempty"`
}

// PlaybackSeekBody is the payload for playback.seek.
type PlaybackSeekBody struct {
	PositionMS int64 `json:"positionMs"`
}

// PlaybackSetVolumeBody is the payload for playback.setVolume.
type PlaybackSetVolumeBody struct {
	Volume float64 `json:"volume"`
}

// PlaybackSetMuteBody is the payload for playback.setMute.
type PlaybackSetMuteBody struct {
	Mute bool `json:"mute"`
}

// QueueGetBody fetches queue entries.
type QueueGetBody struct {
	From    int64  `json:"from"`
	Count   int64  `json:"count"`
	Resolve string `json:"resolve,omitempty"`
}

// QueueGetReply is the reply body for queue.get.
type QueueGetReply struct {
	Revision int64        `json:"revision"`
	Index    int64        `json:"index"`
	Entries  []QueueEntry `json:"entries"`
}

// QueueEntry is the canonical wire representation of a queue position.
// Either Ref or Resolved (or both) must be set. Display is a denormalized
// UI snapshot included whenever known so controllers do not need to issue
// follow-up library catalog lookups for ordinary rendering.
type QueueEntry struct {
	QueueEntryID string           `json:"queueEntryId,omitempty"`
	Ref          *LibraryItemRef  `json:"ref,omitempty"`
	Resolved     *ResolvedSource  `json:"resolved,omitempty"`
	Display      *DisplayMetadata `json:"display,omitempty"`
}

// Validate reports whether a QueueEntry is well formed for wire transport.
func (e QueueEntry) Validate() error {
	if e.Ref == nil && e.Resolved == nil {
		return errors.New("queue entry requires ref or resolved")
	}
	if e.Ref != nil {
		if err := e.Ref.Validate(); err != nil {
			return fmt.Errorf("queue entry ref: %w", err)
		}
	}
	if e.Resolved != nil && e.Resolved.URL == "" {
		return errors.New("queue entry resolved.url is required")
	}
	return nil
}

// QueueSetBody replaces the queue atomically.
type QueueSetBody struct {
	StartIndex int64        `json:"startIndex"`
	Entries    []QueueEntry `json:"entries"`
}

// QueueAddBody adds entries to the queue.
type QueueAddBody struct {
	Position string       `json:"position"`
	AtIndex  *int64       `json:"atIndex,omitempty"`
	Entries  []QueueEntry `json:"entries"`
}

// QueueRemoveBody removes an entry by id or index.
type QueueRemoveBody struct {
	QueueEntryID string `json:"queueEntryId,omitempty"`
	Index        *int64 `json:"index,omitempty"`
}

// QueueMoveBody moves a queue entry.
type QueueMoveBody struct {
	FromIndex int64 `json:"fromIndex"`
	ToIndex   int64 `json:"toIndex"`
}

// QueueJumpBody jumps to an index.
type QueueJumpBody struct {
	Index int64 `json:"index"`
}

// QueueShuffleBody shuffles the queue.
type QueueShuffleBody struct {
	Seed int64 `json:"seed"`
}

// QueueSetShuffleBody toggles shuffle mode.
type QueueSetShuffleBody struct {
	Shuffle bool `json:"shuffle"`
}

// QueueRepeatBody sets repeat mode.
type QueueRepeatBody struct {
	Repeat bool   `json:"repeat"`
	Mode   string `json:"mode,omitempty"`
}

// QueueLoadPlaylistBody loads a playlist into the queue.
type QueueLoadPlaylistBody struct {
	PlaylistServerID string `json:"playlistServerId"`
	PlaylistID       string `json:"playlistId"`
	Mode             string `json:"mode"`
	Resolve          string `json:"resolve"`
}

// QueueLoadSnapshotBody loads a snapshot into the queue.
type QueueLoadSnapshotBody struct {
	PlaylistServerID string `json:"playlistServerId"`
	SnapshotID       string `json:"snapshotId"`
	Mode             string `json:"mode"`
	Resolve          string `json:"resolve"`
}

// QueueLoadSuggestionBody loads a suggestion into the queue.
type QueueLoadSuggestionBody struct {
	PlaylistServerID string `json:"playlistServerId"`
	SuggestionID     string `json:"suggestionId"`
	Mode             string `json:"mode"`
	Resolve          string `json:"resolve"`
}

// ResolvedSource is a fully resolved playable source.
type ResolvedSource struct {
	URL       string `json:"url"`
	Mime      string `json:"mime,omitempty"`
	ByteRange bool   `json:"byteRange"`
}

// PlaylistListBody is the payload for playlist.list.
type PlaylistListBody struct {
	Owner string `json:"owner"`
}

// PlaylistListReply is the reply for playlist.list.
type PlaylistListReply struct {
	Playlists []PlaylistSummary `json:"playlists"`
}

// PlaylistSummary is a playlist summary.
type PlaylistSummary struct {
	PlaylistID string `json:"playlistId"`
	Name       string `json:"name"`
	Revision   int64  `json:"revision"`
}

// PlaylistCreateBody is the payload for playlist.create.
type PlaylistCreateBody struct {
	Name       string `json:"name"`
	SnapshotID string `json:"snapshotId,omitempty"`
}

// PlaylistGetBody is the payload for playlist.get.
type PlaylistGetBody struct {
	PlaylistID string `json:"playlistId"`
}

// PlaylistAddItemsBody is the payload for playlist.addItems.
type PlaylistAddItemsBody struct {
	PlaylistID string       `json:"playlistId"`
	Entries    []QueueEntry `json:"entries"`
}

// PlaylistRemoveItemsBody is the payload for playlist.removeItems.
type PlaylistRemoveItemsBody struct {
	PlaylistID string   `json:"playlistId"`
	EntryIDs   []string `json:"entryIds"`
}

// PlaylistRenameBody is the payload for playlist.rename.
type PlaylistRenameBody struct {
	PlaylistID string `json:"playlistId"`
	Name       string `json:"name"`
}

// PlaylistDeleteBody is the payload for playlist.delete.
type PlaylistDeleteBody struct {
	PlaylistID string `json:"playlistId"`
}

// PlaylistReplaceItemsBody is the payload for playlist.replaceItems.
type PlaylistReplaceItemsBody struct {
	PlaylistID string       `json:"playlistId"`
	Entries    []QueueEntry `json:"entries"`
}

// SnapshotSaveBody is the payload for snapshot.save.
type SnapshotSaveBody struct {
	Name       string          `json:"name"`
	RendererID string          `json:"rendererId"`
	SessionID  string          `json:"sessionId"`
	Capture    SnapshotCapture `json:"capture"`
	Entries    []QueueEntry    `json:"entries,omitempty"`
}

// SnapshotCapture is the session queue capture.
type SnapshotCapture struct {
	QueueRevision int64  `json:"queueRevision"`
	Index         int64  `json:"index"`
	PositionMS    int64  `json:"positionMs"`
	Repeat        bool   `json:"repeat"`
	RepeatMode    string `json:"repeatMode,omitempty"`
	Shuffle       bool   `json:"shuffle"`
}

// SnapshotListBody is the payload for snapshot.list.
type SnapshotListBody struct {
	Owner string `json:"owner,omitempty"`
}

// SnapshotListReply is the reply for snapshot.list.
type SnapshotListReply struct {
	Snapshots []SnapshotSummary `json:"snapshots"`
}

// SnapshotRemoveBody is the payload for snapshot.remove.
type SnapshotRemoveBody struct {
	SnapshotID string `json:"snapshotId"`
}

// SnapshotGetBody is the payload for snapshot.get.
type SnapshotGetBody struct {
	SnapshotID string `json:"snapshotId"`
}

// SnapshotGetReply is the reply for snapshot.get.
type SnapshotGetReply struct {
	SnapshotID string          `json:"snapshotId"`
	Name       string          `json:"name"`
	Revision   int64           `json:"revision"`
	Entries    []QueueEntry    `json:"entries,omitempty"`
	Capture    SnapshotCapture `json:"capture"`
}

// SnapshotSummary describes a saved snapshot.
type SnapshotSummary struct {
	SnapshotID string `json:"snapshotId"`
	Name       string `json:"name"`
	Revision   int64  `json:"revision,omitempty"`
}

// LibraryBrowseBody is the payload for library.browse.
type LibraryBrowseBody struct {
	ContainerID string `json:"containerId"`
	Start       int64  `json:"start"`
	Count       int64  `json:"count"`
}

// LibrarySearchBody is the payload for library.search.
type LibrarySearchBody struct {
	Query string   `json:"query"`
	Start int64    `json:"start"`
	Count int64    `json:"count"`
	Types []string `json:"types,omitempty"`
}

// LibraryGetItemBody is the payload for library.getItem (catalog metadata only).
type LibraryGetItemBody struct {
	Ref LibraryItemRef `json:"ref"`
}

// LibraryGetItemReply describes the response for library.getItem.
type LibraryGetItemReply struct {
	Ref        LibraryItemRef   `json:"ref"`
	Display    *DisplayMetadata `json:"display,omitempty"`
	Attributes map[string]any   `json:"attributes,omitempty"`
}

// LibraryGetItemsBody is the batch payload for library.getItems.
type LibraryGetItemsBody struct {
	Refs []LibraryItemRef `json:"refs"`
}

// LibraryGetItemsReplyItem is one entry in a library.getItems reply.
type LibraryGetItemsReplyItem struct {
	Ref        LibraryItemRef   `json:"ref"`
	Display    *DisplayMetadata `json:"display,omitempty"`
	Attributes map[string]any   `json:"attributes,omitempty"`
	Err        *ReplyError      `json:"err,omitempty"`
}

// LibraryGetItemsReply describes the response for library.getItems.
type LibraryGetItemsReply struct {
	Items []LibraryGetItemsReplyItem `json:"items"`
}

// LibraryResolveSourcesBody is the payload for library.resolveSources (playable urls only).
type LibraryResolveSourcesBody struct {
	Ref LibraryItemRef `json:"ref"`
}

// LibraryResolveSourcesReply describes the response for library.resolveSources.
type LibraryResolveSourcesReply struct {
	Ref     LibraryItemRef   `json:"ref"`
	Sources []ResolvedSource `json:"sources"`
}

// LibraryResolveSourcesBatchBody is the batch payload for library.resolveSourcesBatch.
type LibraryResolveSourcesBatchBody struct {
	Refs []LibraryItemRef `json:"refs"`
}

// LibraryResolveSourcesBatchItem is one entry in a library.resolveSourcesBatch reply.
type LibraryResolveSourcesBatchItem struct {
	Ref     LibraryItemRef   `json:"ref"`
	Sources []ResolvedSource `json:"sources,omitempty"`
	Err     *ReplyError      `json:"err,omitempty"`
}

// LibraryResolveSourcesBatchReply describes the response for library.resolveSourcesBatch.
type LibraryResolveSourcesBatchReply struct {
	Items []LibraryResolveSourcesBatchItem `json:"items"`
}

// SuggestListBody is the payload for suggest.list.
type SuggestListBody struct {
	Owner string `json:"owner,omitempty"`
}

// SuggestListReply is the reply for suggest.list.
type SuggestListReply struct {
	Suggestions []SuggestSummary `json:"suggestions"`
}

// SuggestSummary describes a suggestion.
type SuggestSummary struct {
	SuggestionID string `json:"suggestionId"`
	Name         string `json:"name"`
	Revision     int64  `json:"revision,omitempty"`
}

// SuggestGetBody is the payload for suggest.get.
type SuggestGetBody struct {
	SuggestionID string `json:"suggestionId"`
}

// SuggestPromoteBody is the payload for suggest.promote.
type SuggestPromoteBody struct {
	SuggestionID string `json:"suggestionId"`
	Name         string `json:"name"`
}

File: integrations/android_app/app/src/main/java/com/mediautopia/app/data/protocol/Bodies.kt
package com.mediautopia.app.data.protocol

import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

// --- Session ---

@Serializable
data class SessionAcquireBody(
    val ttlMs: Long,
)

@Serializable
data class SessionRenewBody(
    val ttlMs: Long,
)

@Serializable
data class SessionReplyBody(
    val session: SessionLease,
    val stateVersion: Long,
)

@Serializable
data class SessionLease(
    val id: String,
    val token: String,
    val owner: String,
    val leaseExpiresAt: Long,
)

// --- Playback ---

@Serializable
data class PlaybackPlayBody(
    val index: Long? = null,
)

@Serializable
data class PlaybackSeekBody(
    val positionMs: Long,
)

@Serializable
data class PlaybackSetVolumeBody(
    val volume: Double,
)

@Serializable
data class PlaybackSetMuteBody(
    val mute: Boolean,
)

// --- Library item reference ---

/**
 * Canonical wire reference to a library-backed media item. Encoded as
 * { "kind": "libraryItem", "libraryId": "...", "itemId": "..." }. The
 * server validates that `kind == "libraryItem"`, so we force-emit it
 * even when callers use a Json config without `encodeDefaults = true`.
 */
@OptIn(ExperimentalSerializationApi::class)
@Serializable
data class LibraryItemRef(
    @EncodeDefault val kind: String = LIBRARY_ITEM_KIND,
    val libraryId: String,
    val itemId: String,
) {
    companion object {
        const val LIBRARY_ITEM_KIND = "libraryItem"
    }
}

/**
 * Denormalized UI snapshot carried alongside library refs so controllers
 * can render queue and now-playing without follow-up catalog lookups.
 * All fields are optional; consumers must tolerate missing values.
 */
@Serializable
data class DisplayMetadata(
    val title: String? = null,
    val artist: String? = null,
    val artists: List<String>? = null,
    val album: String? = null,
    val artworkUrl: String? = null,
    val durationMs: Long? = null,
    val mediaType: String? = null,
)

// --- Queue ---

@Serializable
data class QueueGetBody(
    val from: Long,
    val count: Long,
    val resolve: String = "",
)

@Serializable
data class QueueGetReply(
    val revision: Long,
    val index: Long,
    val entries: List<QueueEntry>,
)

@Serializable
data class QueueSetBody(
    val startIndex: Long,
    val entries: List<QueueEntry>,
)

@Serializable
data class QueueAddBody(
    val position: String,
    val atIndex: Long? = null,
    val entries: List<QueueEntry>,
)

@Serializable
data class QueueRemoveBody(
    val queueEntryId: String = "",
    val index: Long? = null,
)

@Serializable
data class QueueMoveBody(
    val fromIndex: Long,
    val toIndex: Long,
)

@Serializable
data class QueueJumpBody(
    val index: Long,
)

@Serializable
data class QueueShuffleBody(
    val seed: Long,
)

@Serializable
data class QueueSetShuffleBody(
    val shuffle: Boolean,
)

@Serializable
data class QueueRepeatBody(
    val repeat: Boolean,
    val mode: String = "",
)

@Serializable
data class QueueLoadPlaylistBody(
    val playlistServerId: String,
    val playlistId: String,
    val mode: String,
    val resolve: String,
)

@Serializable
data class QueueLoadSnapshotBody(
    val playlistServerId: String,
    val snapshotId: String,
    val mode: String,
    val resolve: String,
)

@Serializable
data class QueueLoadSuggestionBody(
    val playlistServerId: String,
    val suggestionId: String,
    val mode: String,
    val resolve: String,
)

@Serializable
data class QueueEntry(
    val queueEntryId: String = "",
    val ref: LibraryItemRef? = null,
    val resolved: ResolvedSource? = null,
    val display: DisplayMetadata? = null,
)

@Serializable
data class ResolvedSource(
    val url: String,
    val mime: String = "",
    val byteRange: Boolean = false,
)

// --- Library ---

@Serializable
data class LibraryBrowseBody(
    val containerId: String,
    val start: Long,
    val count: Long,
)

@Serializable
data class LibrarySearchBody(
    val query: String,
    val start: Long,
    val count: Long,
    val types: List<String> = emptyList(),
)

@Serializable
data class LibraryGetItemBody(
    val ref: LibraryItemRef,
)

@Serializable
data class LibraryGetItemReply(
    val ref: LibraryItemRef,
    val display: DisplayMetadata? = null,
    val attributes: JsonObject? = null,
)

@Serializable
data class LibraryGetItemsBody(
    val refs: List<LibraryItemRef>,
)

@Serializable
data class LibraryGetItemsReplyItem(
    val ref: LibraryItemRef,
    val display: DisplayMetadata? = null,
    val attributes: JsonObject? = null,
    val err: ReplyError? = null,
)

@Serializable
data class LibraryGetItemsReply(
    val items: List<LibraryGetItemsReplyItem>,
)

@Serializable
data class LibraryResolveSourcesBody(
    val ref: LibraryItemRef,
)

@Serializable
data class LibraryResolveSourcesReply(
    val ref: LibraryItemRef,
    val sources: List<ResolvedSource> = emptyList(),
)

@Serializable
data class LibraryResolveSourcesBatchBody(
    val refs: List<LibraryItemRef>,
)

@Serializable
data class LibraryResolveSourcesBatchItem(
    val ref: LibraryItemRef,
    val sources: List<ResolvedSource> = emptyList(),
    val err: ReplyError? = null,
)

@Serializable
data class LibraryResolveSourcesBatchReply(
    val items: List<LibraryResolveSourcesBatchItem>,
)

// --- Playlist ---

@Serializable
data class PlaylistListBody(
    val owner: String,
)

@Serializable
data class PlaylistListReply(
    val playlists: List<PlaylistSummary>,
)

@Serializable
data class PlaylistSummary(
    val playlistId: String,
    val name: String,
    val revision: Long,
)

@Serializable
data class PlaylistCreateBody(
    val name: String,
    val snapshotId: String = "",
)

@Serializable
data class PlaylistGetBody(
    val playlistId: String,
)

@Serializable
data class PlaylistAddItemsBody(
    val playlistId: String,
    val entries: List<QueueEntry>,
)

@Serializable
data class PlaylistRemoveItemsBody(
    val playlistId: String,
    val entryIds: List<String>,
)

@Serializable
data class PlaylistRenameBody(
    val playlistId: String,
    val name: String,
)

@Serializable
data class PlaylistDeleteBody(
    val playlistId: String,
)

@Serializable
data class PlaylistReplaceItemsBody(
    val playlistId: String,
    val entries: List<QueueEntry>,
)

// --- Snapshot ---

@Serializable
data class SnapshotSaveBody(
    val name: String,
    val rendererId: String,
    val sessionId: String,
    val capture: SnapshotCapture,
    val entries: List<QueueEntry> = emptyList(),
)

@Serializable
data class SnapshotCapture(
    val queueRevision: Long,
    val index: Long,
    val positionMs: Long,
    val repeat: Boolean,
    val repeatMode: String = "",
    val shuffle: Boolean,
)

@Serializable
data class SnapshotListBody(
    val owner: String = "",
)

@Serializable
data class SnapshotListReply(
    val snapshots: List<SnapshotSummary>,
)

@Serializable
data class SnapshotRemoveBody(
    val snapshotId: String,
)

@Serializable
data class SnapshotGetBody(
    val snapshotId: String,
)

@Serializable
data class SnapshotGetReply(
    val snapshotId: String,
    val name: String,
    val revision: Long,
    val entries: List<QueueEntry> = emptyList(),
    val capture: SnapshotCapture,
)

@Serializable
data class SnapshotSummary(
    val snapshotId: String,
    val name: String,
    val revision: Long = 0,
)

// --- Suggestion ---

@Serializable
data class SuggestListBody(
    val owner: String = "",
)

@Serializable
data class SuggestListReply(
    val suggestions: List<SuggestSummary>,
)

@Serializable
data class SuggestSummary(
    val suggestionId: String,
    val name: String,
    val revision: Long = 0,
)

@Serializable
data class SuggestGetBody(
    val suggestionId: String,
)

@Serializable
data class SuggestPromoteBody(
    val suggestionId: String,
    val name: String,
)
